File: Tests/AINotebookCoreTests/NoteIndexerTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NoteIndexerTests: XCTestCase {

    func testIndexCreatesShadowSourceAndChunks() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(
            notebookId: nb.id!, title: "Recipe", bodyMd: "Mix flour and water."
        )

        let indexer = NoteIndexer(store: store)
        try await indexer.index(noteId: n.id!)

        let reloaded = try XCTUnwrap(try store.note(id: n.id!))
        let sourceId = try XCTUnwrap(reloaded.autoSourceId)

        let shadow = try XCTUnwrap(try store.source(id: sourceId))
        XCTAssertEqual(shadow.type, .note)
        XCTAssertEqual(shadow.notebookId, nb.id!)
        XCTAssertEqual(shadow.title, "Recipe")
        XCTAssertEqual(shadow.status, .ready)

        let chunks = try store.chunks(sourceId: sourceId)
        XCTAssertGreaterThan(chunks.count, 0)
        XCTAssertTrue(chunks.first?.text.contains("flour") == true)
    }

    func testReindexReplacesChunks() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(
            notebookId: nb.id!, title: "T", bodyMd: "original"
        )
        let indexer = NoteIndexer(store: store)
        try await indexer.index(noteId: n.id!)

        try store.updateNote(id: n.id!, title: "T", bodyMd: "rewritten body completely")
        try await indexer.index(noteId: n.id!)

        let sourceId = try XCTUnwrap(try store.note(id: n.id!)?.autoSourceId)
        let chunks = try store.chunks(sourceId: sourceId)
        XCTAssertEqual(chunks.count, 1)
        XCTAssertTrue(chunks[0].text.contains("rewritten"))
        XCTAssertFalse(chunks[0].text.contains("original"))
    }

    func testEmptyBodyClearsChunksButKeepsShadow() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(
            notebookId: nb.id!, title: "T", bodyMd: "first content"
        )
        let indexer = NoteIndexer(store: store)
        try await indexer.index(noteId: n.id!)

        try store.updateNote(id: n.id!, title: "T", bodyMd: "")
        try await indexer.index(noteId: n.id!)

        let sourceId = try XCTUnwrap(try store.note(id: n.id!)?.autoSourceId)
        XCTAssertNotNil(try store.source(id: sourceId))
        XCTAssertEqual(try store.chunks(sourceId: sourceId).count, 0)
    }

    func testKickHookFiresAfterIndex() async throws {
        final class CapturingKick: @unchecked Sendable {
            let lock = NSLock()
            var count = 0
            func bump() { lock.lock(); count += 1; lock.unlock() }
            var value: Int { lock.lock(); defer { lock.unlock() }; return count }
        }
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let n = try store.createNote(
            notebookId: nb.id!, title: "T", bodyMd: "x"
        )
        let kick = CapturingKick()
        let indexer = NoteIndexer(store: store, onChunksWritten: { kick.bump() })
        try await indexer.index(noteId: n.id!)
        XCTAssertEqual(kick.value, 1)
    }
}

File: windows/tests/AINotebook.Core.Tests/Rag/NoteIndexerTests.cs
using AINotebook.Core;
using AINotebook.Core.Models;
using AINotebook.Core.Rag;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Rag;

public class NoteIndexerTests
{
    // NoteIndexerTests.testIndexCreatesShadowSourceAndChunks
    [Fact]
    public async Task IndexCreatesShadowSourceAndChunks()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var note = store.CreateNote(nb.Id!.Value, "Recipe", "Mix flour and water.", NoteOrigin.Manual, null);

        var indexer = new NoteIndexer(store);
        await indexer.IndexAsync(note.Id!.Value);

        var refreshed = store.Note(note.Id!.Value)!;
        Assert.NotNull(refreshed.AutoSourceId);
        var shadow = store.Source(refreshed.AutoSourceId!.Value)!;
        Assert.Equal(SourceType.Note, shadow.Type);
        Assert.Equal("Recipe", shadow.Title);
        Assert.Equal(SourceStatus.Ready, shadow.Status);

        var chunks = store.Chunks(shadow.Id!.Value);
        Assert.NotEmpty(chunks);
        Assert.Contains(chunks, c => c.Text.Contains("flour"));
    }

    // NoteIndexerTests.testReindexReplacesChunks
    [Fact]
    public async Task ReindexReplacesChunks()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var note = store.CreateNote(nb.Id!.Value, "T", "original body", NoteOrigin.Manual, null);
        var indexer = new NoteIndexer(store);
        await indexer.IndexAsync(note.Id!.Value);

        store.UpdateNote(note.Id!.Value, "T", "replaced body");
        await indexer.IndexAsync(note.Id!.Value);

        var shadow = store.Source(store.Note(note.Id!.Value)!.AutoSourceId!.Value)!;
        var chunks = store.Chunks(shadow.Id!.Value);
        Assert.Contains(chunks, c => c.Text.Contains("replaced"));
        Assert.DoesNotContain(chunks, c => c.Text.Contains("original"));
    }

    // NoteIndexerTests.testEmptyBodyClearsChunksButKeepsShadow
    [Fact]
    public async Task EmptyBodyClearsChunksButKeepsShadow()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var note = store.CreateNote(nb.Id!.Value, "T", "has content", NoteOrigin.Manual, null);
        var indexer = new NoteIndexer(store);
        await indexer.IndexAsync(note.Id!.Value);

        store.UpdateNote(note.Id!.Value, "T", "   ");
        await indexer.IndexAsync(note.Id!.Value);

        var shadowId = store.Note(note.Id!.Value)!.AutoSourceId!.Value;
        Assert.NotNull(store.Source(shadowId));
        Assert.Empty(store.Chunks(shadowId));
    }

    // NoteIndexerTests.testKickHookFiresAfterIndex
    [Fact]
    public async Task KickHookFiresAfterIndex()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var note = store.CreateNote(nb.Id!.Value, "T", "body", NoteOrigin.Manual, null);

        var fired = 0;
        var indexer = new NoteIndexer(store, () => { fired++; return Task.CompletedTask; });
        await indexer.IndexAsync(note.Id!.Value);

        Assert.Equal(1, fired);
    }
}
